### PR TITLE
Add error handling for decoding errors

### PR DESCRIPTION
**Pull request by mrbean-bremen**
- use log collector for unit tests
- see #1200

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- this adds only warnings about decoding - some warnings about incorrect Specific Character Set values are still missing, I may add a separate PR for these later
- I added a logger that collects log messages and currently use it in all tests - I failed to use it only for a certain test collection (probably missing some understanding here, haven't used xUnit before)
- I use the logger as proposed, not sure if there is a better way

### CODE REVIEW

File: FO-DICOM.Core/DicomEncoding.cs [hunk 1]
```diff
@@ -60,52 +61,66 @@ public static Encoding GetEncoding(string charset)
...
         /// </summary>
         /// <param name="encoding">Encoding.</param>
-        /// <returns>Charset.</returns>
-        public static string GetCharset(Encoding encoding) =>
-            // Do we always want the extended charsets?
-            encoding?.WebName switch
+        /// <param name="extended">If true, the extended version of the character set is returned.</param>
+        /// <returns>The Specific Character Set as defined in DICOM.</returns>
+        /// <exception cref="ArgumentException">No character set found for the encoding.</exception>
+        public static string GetCharset(Encoding encoding, bool extended = false)
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/DicomEncoding.cs:113]: This is actually unrelated to the issue, just noticed that the method did not support all encodings and re-wrote it to be generic. It seems only to be used in tests.

File: FO-DICOM.Core/Log/ConsoleLogger.cs [hunk 2]
```diff
@@ -21,7 +21,7 @@ public class ConsoleLogger : TextWriterLogger
         /// <summary>
         /// Initializes an instance of the <see cref="ConsoleLogger"/>.
         /// </summary>
-        private ConsoleLogger() : base(Console.Out)
+        protected ConsoleLogger() : base(Console.Out)
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/Log/ConsoleLogger.cs:24]: I derive from the ConsoleLogger.

File: Tests/FO-DICOM.Tests/CollectingConsoleLogger.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+using System.Linq;
+using FellowOakDicom.Log;
+
+namespace FellowOakDicom.Tests
+{
+    /// <summary>
+    /// Console Logger that additionally collects the log messages for use in unit tests.
+    /// </summary>
+    class CollectingConsoleLogger : ConsoleLogger
+    {
```
Comment (mrbean-bremen) [inline Tests/FO-DICOM.Tests/CollectingConsoleLogger.cs:14]: I derive from the console logger to have the logging done to the console during the tests as before - not sure if this is actually needed.
Also, maybe there is already a logger that does that in xUnit - I didn't find one, but I may have missed it.

File: Tests/FO-DICOM.Tests/CollectingConsoleLogger.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        /// <param name="logLevel">Only messages of this log level are considered.</param>
+        /// </summary>
+        public string LogMessageAt(int index, LogLevel logLevel) =>
+            _logEntries.Where(e => e.Item1 == logLevel).ElementAt(index).Item2;
+
+        /// <summary>
+        /// Returns the warning message at a given index.
+        /// <param name="index">The index of the message in all warning messages.</param>
+        /// </summary>
+        public string WarningAt(int index) => LogMessageAt(index, LogLevel.Warning);
```
Comment (mrbean-bremen) [inline Tests/FO-DICOM.Tests/CollectingConsoleLogger.cs:64]: I added only a couple of methods that I needed, more can be added later.

File: Tests/FO-DICOM.Tests/DicomEncodingTest.cs [hunk 5]
```diff
@@ -54,25 +60,38 @@ public void FallbackEncoding()
         {
             var ds = new DicomDataset
             {
-                { DicomTag.SOPClassUID, DicomUID.SecondaryCaptureImageStorage },
-                { DicomTag.SOPInstanceUID, "1.2.345.67890" }
+                {DicomTag.SOPClassUID, DicomUID.SecondaryCaptureImageStorage},
+                {DicomTag.SOPInstanceUID, "1.2.345.67890"}
```
Comment (mrbean-bremen) [inline Tests/FO-DICOM.Tests/DicomEncodingTest.cs:64]: I only just noticed that the autoformat removed these spaces - I can revert this if, of course.

File: Tests/FO-DICOM.Tests/Fixture.cs [hunk 6]
```diff
@@ -9,13 +9,18 @@
...
+    /// to allow testing log messages.
+    /// </summary>
+    public class GlobalFixture: IDisposable
     {
 
         public GlobalFixture()
         {
             var serviceCollection = new ServiceCollection()
-                .AddFellowOakDicom();
+                .AddFellowOakDicom().AddLogManager<CollectingConsoleLogManager>();
```
Comment (mrbean-bremen) [inline Tests/FO-DICOM.Tests/Fixture.cs:23]: I made a separate fixture to use the collecting log manager first, added a collection definition and used it in the DicomEncodingTest, but it turned out that this does not work if running several test collections. In this case, only the first test collection was used. I tried to debug this, but in the debugger it behaves as I would have expected.
Maybe someone can help me with that... For the time being, I just changed the logger for all tests, that may generate a bit of an unneded overhead.

**Comment by mrbean-bremen** [comment]
Just to clarify - this is ready for review.

**Comment by gofal** [comment]
Thanks! I have been on vacation, but now are going to review and merge.

**Comment by mrbean-bremen** [comment]
Take your time :)

File: FO-DICOM.Core/DicomEncoding.cs (hunk 1 above)
Comment (gofal) [reply FO-DICOM.Core/DicomEncoding.cs:113]: within fo-dicom this method is only used in tests. The intention is: when a developer using fo-dicom creates a dataset then he has to set the SpecificCharacterSet. This method helps to get the character set if you have the System.Text.Encoding given (e.g. set by operating system or by configuration). I have no idea, how many applications out there actually are using this methods or if any.

File: Tests/FO-DICOM.Tests/CollectingConsoleLogger.cs (hunk 3 above)
Comment (gofal) [reply Tests/FO-DICOM.Tests/CollectingConsoleLogger.cs:14]: xunit has some classes to capture log output, it is done with the ITestOutputHelper class, that is wrapped in fo-dicom.Tests with XUnitDicomLogger (deriving from ILogger). But with that class you cannot access the output and assert some log messages. So your approach is just fine for testing log output!

File: Tests/FO-DICOM.Tests/Fixture.cs (hunk 6 above)
Comment (gofal) [reply Tests/FO-DICOM.Tests/Fixture.cs:23]: The issue is the static property DicomEncoding.Logger. when xunit runs the collections one after the other, then on the first call to this property an instance is resolved from the service provider and stored in the static property. But on the next collection the value of the static property is still set and therefore will never be resolved from the new service provider. To make the DicomEncoding class work fine with various service providers, then the DicomEncoding class should not store the ILogger instance in a static field but resolve it from ServiceProvider everytime it needs it. This may also be fine, because resolving a service should not be too long and the DicomEncoding class only does logging in case of some Warnings.

File: FO-DICOM.Core/DicomEncoding.cs [hunk 7]
```diff
@@ -289,18 +317,45 @@ private static string DecodeFragment(byte[] fragment, Encoding[] encodings, byte
             }
             else
             {
-                for (int i = seqLength; i < fragment.Length; i++)
+                for (int i = seqLength; i < fragment.Length - 1; i++)
```
Comment (gofal) [inline FO-DICOM.Core/DicomEncoding.cs:320]: is this intended, that the loop previously iterated all fragement and now all but the last one?

File: FO-DICOM.Core/DicomEncoding.cs [hunk 8]
```diff
@@ -60,91 +61,109 @@ public static Encoding GetEncoding(string charset)
...
+            { "ISO 2022 IR 58", "gb2312" }, // Chinese (Simplified) Extended
+            { "ISO 2022 GBK", "GBK" }, // Chinese (Simplified) Extended (supersedes GB2312)
         };
 
-        private static readonly IDictionary<string, Encoding> _knownEncodings = new Dictionary<string, Encoding>();
-        
+        private static readonly IDictionary<string, Encoding> _knownEncodings =
+            new Dictionary<string, Encoding>();
+
+        private static readonly IDictionary<int, Encoding> _strictEncodings =
```
Comment (gofal) [inline FO-DICOM.Core/DicomEncoding.cs:164]: Whats the difference between _knownEncodings and _strictEncodings?

File: Tests/FO-DICOM.Tests/Fixture.cs (hunk 6 above)
Comment (mrbean-bremen) [reply Tests/FO-DICOM.Tests/Fixture.cs:23]: Ah ok - should have seen that! I will change it as proposed.

File: FO-DICOM.Core/DicomEncoding.cs (hunk 7 above)
Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:320]: Yes, I think this was actually a bug - will have a closer look later today to make sure this is correct.

File: FO-DICOM.Core/DicomEncoding.cs (hunk 8 above)
Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:164]: `_strictEncodings` raise on error (to be able to actually detect encoding/decoding errors), `knowEncodings` use replacement characters on error. Will add some comments, and probably adapt the names.

File: FO-DICOM.Core/DicomEncoding.cs (hunk 7 above)
Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:320]: Ok, there actually _is_ a bug. The bug is not that the last character is not checked (this is not needed, because even if it is delimiter, there is nothing to convert after it), but in the next line, where the passed length is incorrect (and if including the last character, some tests will fail for that reason). I have been sloppy here...

File: FO-DICOM.Core/DicomEncoding.cs (hunk 8 above)
Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:171]: Hm, I added a description, but couldn't find good names that are not ridiculously long... maybe you have an idea, I often struggle with names

Comment (gofal) [reply FO-DICOM.Core/DicomEncoding.cs:171]: welcome to the club....

**Comment by gofal** [comment]
The Setup.ServiceProvider is a static property, and xunit executes the collections in parallel (different threads). So solve this, there is a TestServiceProviderHost, that works like a ServiceProvider, but it has internally a dictionary of IServiceProviders. When a service is requested, then this TestServiceProviderHost checks callstack for the collection and returns a service out of the matching IServiceProvider. So the right way here in fo-dicom is not to create separate fixture classes, but to create a new serviceCollection in constructor of GlobalFixture and to register this in TestServiceProviderHost.

File: Tests/FO-DICOM.Tests/Fixture.cs (hunk 6 above)
Comment (mrbean-bremen) [reply Tests/FO-DICOM.Tests/Fixture.cs]: Ok, not making the logger a singleton did not resolve the issue. I see that you have reverted to use the same collection for all tests as before - I'm still not sure why having more than one collection class did not work.

File: FO-DICOM.Core/DicomEncoding.cs (hunk 7 above)
Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:327]: Shall be fixed now.

File: FO-DICOM.Core/DicomEncoding.cs (hunk 1 above)
Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:113]: Ok, in this case I think it is better to return the normal version (instead of the extended) of the encoding as is done now by default, otherwise `SpecificCharacterSet` would not be valid in the vast majority of single-value encodings. I doubt that it has been used though, as somebody would have noticed that...
